src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,573 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <boost/iterator/counting_iterator.hpp>
+
+#include "include/intarith.h"
+#include "crimson/os/seastore/journal/circular_bounded_journal.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore {
+
+std::ostream &operator<<(std::ostream &out, const CBJournal::cbj_header_t &header)
+{
+  return out << "cbj_header_t(magin=" << header.magic
+	     << ", uuid=" << header.uuid
+	     << ", block_size=" << header.block_size
+	     << ", size=" << header.size
+	     << ", used_size=" << header.used_size
+	     << ", error=" << header.error
+	     << ", start_offset=" << header.start_offset
+	     << ", applied_to="<< header.applied_to
+	     << ", last_committed_record_base="<< header.last_committed_record_base
+	     << ", written_to=" << header.written_to
+	     << ", flsg=" << header.flag
+	     << ", csum_type=" << header.csum_type
+	     << ", csum=" << header.csum
+	     << ", start=" << header.start
+	     << ", end=" << header.end
+             << ")";
+}
+
+
+CBJournal::CBJournal(NVMeBlockDevice* device, const std::string path)
+  : device(device), path(path) {}
+
+CBJournal::mkfs_ret
+CBJournal::mkfs(mkfs_config_t& config)
+{
+  return _open_device(path
+  ).safe_then([this, &config]() {
+    rbm_abs_addr start_addr = convert_paddr_to_abs_addr(
+      config.start);
+    assert(config.block_size == device->get_block_size());
+    return read_super(start_addr
+    ).safe_then([](auto super) {
+      return mkfs_ertr::now();
+    }).handle_error(
+      crimson::ct_error::enoent::handle([this, &config, start_addr] (auto) 
+	-> mkfs_ret {
+	ceph::bufferlist bl;
+	CBJournal::cbj_header_t head;
+	try {
+	  head.magic = CBJOURNAL_MAGIC;
+	  head.uuid = uuid_d(); // TODO
+	  head.block_size = config.block_size;
+	  rbm_abs_addr end_addr = convert_paddr_to_abs_addr(
+	    config.end);
+	  head.size = end_addr - start_addr
+	    - device->get_block_size();
+	  head.used_size = 0;
+	  head.error = 0;
+	  head.start_offset = device->get_block_size();
+	  head.last_committed_record_base = 0;
+	  head.written_to = head.start_offset;
+	  head.applied_to = 0;
+	  head.flag = 0;
+	  head.csum_type = 0;
+	  head.csum = 0;
+	  head.cur_segment_seq = 0;
+	  head.start = start_addr;
+	  head.end = end_addr;
+	  head.device_id = config.device_id;
+	  encode(head, bl);
+	} catch (ceph::buffer::error &e) {
+	  logger().debug("unable to encode super block from underlying deivce");
+	  return mkfs_ertr::now();
```

Comment:
Is this pathway actually possible?  This can't happen unless the encode implementation is actually broken, I don't think returning success is the right answer if so.